In [1]:
import pandas as pd

In [2]:
card=pd.read_csv(r"../BankData/card.csv", sep=";")
card.head()

,card_id,disp_id,type,issued
0,1005,9285,classic,931107 00:00:00
1,104,588,classic,940119 00:00:00
2,747,4915,classic,940205 00:00:00
3,70,439,classic,940208 00:00:00
4,577,3687,classic,940215 00:00:00


In [3]:
card.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   card_id  892 non-null    int64 
 1   disp_id  892 non-null    int64 
 2   type     892 non-null    object
 3   issued   892 non-null    object
dtypes: int64(2), object(2)
memory usage: 28.0+ KB


In [4]:
card.duplicated().sum()

0

In [5]:
card["card_id"].nunique()

892

In [6]:
card["type"].value_counts()

type
classic    659
junior     145
gold        88
Name: count, dtype: int64

In [7]:
card["card_id"] = card["card_id"].astype(str)
card["disp_id"] = card["disp_id"].astype(str)

In [8]:
card["issued"] = card["issued"].astype(str).str.split(" ").str[0]
card["issued"] = pd.to_datetime(card["issued"].str.zfill(6), format="%y%m%d")

In [9]:
card.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   card_id  892 non-null    object        
 1   disp_id  892 non-null    object        
 2   type     892 non-null    object        
 3   issued   892 non-null    datetime64[ns]
dtypes: datetime64[ns](1), object(3)
memory usage: 28.0+ KB


In [10]:
card.head()

,card_id,disp_id,type,issued
0,1005,9285,classic,1993-11-07
1,104,588,classic,1994-01-19
2,747,4915,classic,1994-02-05
3,70,439,classic,1994-02-08
4,577,3687,classic,1994-02-15


## Stage 5: Data Cleaning & Transformation — `card` Table

### 1. Data Cleaning & Transformation Log

| Column Name | Raw Type | Cleaned Type | Original / Raw Value | Converted / Mapped Value | Business / Risk Rationale |
| :--- | :---: | :---: | :--- | :--- | :--- |
| **`card_id`** | `int64` | `str` | Numeric integer (e.g., `1005`) | String object (e.g., `'1005'`) | Primary key. Converted to string to prevent numerical math operations and ensure pipeline consistency. |
| **`disp_id`** | `int64` | `str` | Numeric integer (e.g., `9285`) | String object (e.g., `'9285'`) | Foreign key linking to the `disp` (disposition/relationship) table. |
| **`type`** | `str` | `str` | Clean string (`classic`, `junior`, `gold`) | Retained as-is (Validated) | Validated via `value_counts()`. No transformation required; used for cross-sell modeling. |
| **`issued`** | `str` / `int64` | `datetime64[ns]` | YYMMDD timestamp string (e.g., `931107 00:00:00`) | ISO standard datetime (e.g., `1993-11-07`) | Represents card issue date. Extracted date substring, zero-padded via `.str.zfill(6)`, and converted to datetime. |

---

### 2. Validation & Integrity Audit Checklist

- [x] **Null Values:** 0 missing values detected across all columns.
- [x] **Primary Key Uniqueness:** `card_id` contains 892 unique values (100% unique, zero duplicates).
- [x] **Date Conversion:** Successfully converted `issued` to standard `YYYY-MM-DD` timestamps.
- [x] **Categorical Consistency:** `type` strictly contains valid card tiers (`classic`: 659, `junior`: 145, `gold`: 88).

In [11]:
card.to_csv("../cleanData/clean_card.csv", sep=";", index=False)